Running deployments of cluster-manager and klusterlet outside of Hub and ManagedCluster (#36)
open-cluster-management-io/enhancements: enhancements/sig-architecture/33-detached-deploy-mode/detached-deploy-mode.md

        +------------------------v--------------------------------+                                               
        |  hub-cluster                                            |       registrater       +-----------------+   
        |                                                         <-------------------------> managed cluster |   
        |  +-------+ +-----------+ +---------------+              |                         +-----------------+   
        |  | CRDs  | |APIServices| |Configurations | etc.         |                                               
        |  |       | |           | |               |              |                                               
        |  +-------+ +-----------+ +---------------+              |                                               
        +---------------------------------------------------------+
```

The **hub-cluster** is where hub is installed and where the cluster agent should register.
 
The **management-cluster** is the cluster where deployments are actually running on. The deployments of several hub-clusters could run in one management-cluster and be divided by different namespaces.

> Note that cluster-manager is deploy on **management-cluster** but configed with a kubeconfig secret (in the `open-cluster-management` namespace) named ”<cluster-manager-name>-hub-kubeconfig” which is a cluster-admin kubeconfig of the **hub-cluster**.

Reviewer: `<cluster-manager-name>-hub-kubeconfig` means for the external hub kubeconfig? Please confirm it is always deployed in the `open-cluster-management`?

Author: Yes, `<cluster-manager-name>-hub-kubeconfig` means the external hub kubeconfig.  And the current approach is all of cluster-manager's external-hub-kubeconfig will be depolyed in the `open-cluster-management` namespace.

Reviewer: Users create `<cluster-manager-name>-hub-kubeconfig` in `open-cluster-management` ns, then the operator copy it to the installation ns?

Author: It could be: operator use this <cluster-manager-name>-hub-kubeconfig to generate proper SA-kubeconfig and install them in respective installation ns.

Author: Solved, apply secret after namespace is created

|                                                                  |                                              
|  +----------------------+   +----------------------+             |                                              
|  | CRDs:                |   | Configurations:      |             |                                              
|  | clusterClaim         |   | role,rolebinding,sa  |   etc.      |                                              
|  | appliedmanifestwork  |   |                      |             |                                              
|  +----------------------+   +----------------------+             |                                              
+------------------------------------------------------------------+  
```

- The managed-cluster is the cluster managed by the hub cluster, manifestwork will be deployed onto this cluster.
- The management-cluster is the cluster where Klusterlet related deployments are actually running.

The managed cluster will not connect to the hub cluster, all things done by the management cluster. The management cluster could be the hub cluster, it could be another third-party cluster as well.

> In Detached mode, the "klusterlet.spec.namespace" will be ignored, all agent components will be installed to the namespace named <klusterlet's name>-open-cluster-management-agent. The other thing to note is that users are responsible for creating an external-managed-kubeconfig in the agent installation namespace, Klusterlet-operator will use this kubeconfig to get a new minimum permission kubeconfig for the managed cluster, then render it to registration-agent and work-agent.

Reviewer: ```suggestion
In Detached mode, the "klusterlet.spec.namespace" will be ignored, all agent components will be installed to the namespace with the same name as klusterlet. The other thing to note is that users are responsible for creating an `external-managed-kubeconfig` in the agent installation namespace, Klusterlet-operator will use this kubeconfig to get a new minimum permission kubeconfig for the managed cluster, then render it to registration-agent and work-agent.
```

Author: The commit is squshed to and sign information.

                            |                                              |                                            
                   watch,create,update,delete...            watch,create,update,delete...                               
                            |                                              |                                            
                    +-------v-----------+                     +------------v----------+                                 
                    |hub                |                     |managed cluster        |                                 
                    |                   |     logically       |                       |                                 
                    | +--------+        |<------------------->| +--------+            |                                 
                    | | CRDs...|        |                     | |CRDs... |            |                                 
                    | +--------+        |                     | +--------+            |                                 
                    +-------------------+                     +-----------------------+     
```

> A real use case would be: The hub cluster is installed in **private network** while the managed cluster is in public cloud. right now, the managed cluster cannot connect to the hub cluster. **It also means deployments on managed cluster can not apply any resources to the hub cluster directly**.

Author: @clyang82  Hello, Chunlin, we have added a real use case here, please have a review and see whether it's clear for you. Thanks~

Author: @qiujian16 The use case is added now, please also have a review, thanks~

Author: Updated this part, the link has outdated.

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [website](https://github.com/open-cluster-management/website/)

## Summary

The proposed work would support running deployments of cluster-manager and klusterlet outside of Hub and ManagedCluster. 

## Motivation

Currently, the cluster manager and klusterlet should be deployed on the hub/managed cluster, it is mandatory to expose the hub cluster to the managed cluster even if the managed cluster resides in an insecure environment. 

To address this, we propose to provide an option for running cluster-manager/klusterlet outside of the hub/managed cluster.

This ability will bring some additional possible benefits:
* We can leverage compute resources better by running several cluster-managers/klusterlets deployments on the same powerful cluster; it also enhances our management of all deployments.
* The footprints of the hub/managed cluster can be reduced.

> * We can use this ability to run the klusterlet on the hub cluster, so we do not need to expose the hub cluster to the managed cluster, and accordingly, we expose the managed cluster to the hub (A bit like Pull Mode to Push Mode).

Reviewer 2: I wonder if this approach makes the management cluster a performance bottleneck? What is the advantage of running Klusterlets in a management cluster vs a pure push model?

Author: We had a case that a customer has a super-powerful machine and want to use it to run every deployment of each hub cluster. So base on this situation, the management cluster would be a bottleneck, but if the situation changes, we might need to use another approach. 
For the advantage, I would say the compatibility since there are no API changes. @zhujian7  Feel free to add more.
And CSR is still needed for this.

Reviewer 3: @xuezhaojun I think it worth writing a section to compare it with push mode (cons and pros). I think a push mode will need listwatch on multiple spokes, which has been proven to be not scalable in kubefed since a single controller will consume a lot of memory. Running klusterlet separated distribute the computation to multiple pods. It might be good to consider even in push case, we still have workers for each different spoke.

Author: The content added.

Reviewer 2: good point, the push model shouldn't require to run all of them in the same pod.

### Goals

* Support deploy cluster-manager outside of the hub cluster.
* Support deploy klusterlet outside of the managed cluster.

### Non-Goals

* Support deploy addons agents outside of the managed cluster.(TBD)
* Authenticate managed cluster without using CSR(CertificateSigningRequest).

## User Stories

### Story1

As a managed cluster administrator, I hope as few as possible workloads running in my cluster while the cluster can be managed by the hub.

### Story2

> As a hub cluster administrator, on the premise that an untrusted cluster can be managed, I do not want the cluster to talk to the hub cluster directly.

Reviewer 2: This seems to be a call for a push model

Author: Yes, that's right. It could be solved by a push model as well.

### Story3

> As an administrator of several Hubs, I want to run all deployments in one place for a better management experience,

Reviewer 2: IIRC, this is a fleet of a fleet case. Would it make sense to create a uber hub cluster to manage all the hub clusters as member clusters instead?

Author: In this case, we aim to do no management. Instead, for example,  we can easily collect logs, metrics, and other data of running pods from all hubs.

Reviewer 3: this is interesting. We had discussion before on such uber hub idea. The thing that is not clear is what scenario want to achieve with this uber hub. Do we just want to provide a index so the request can be proxied to a certain second level hub or we want a higher level coordination, or just checking the status of the clusters. I think scalability also needs to be considered with this uber hub. a single can support manage than 2k cluster today. A uber hub might need to managed 10/100 times more.

                            |                      |  <public cloud>       |                                            
                   watch,create,update,delete...   |         watch,create,update,delete...                               
                            |                      |                       |                                            
                    +-------v-----------+          |           +-----------v-----------+                                 
                    |hub                |          |           |managed cluster        |                                 
                    |                   |     logic|ally       |                       |                                 
                    | +--------+        |<- - - - -|- - - - - >| +--------+            |                                 
                    | | CRDs...|        |          |           | |CRDs... |            |                                 
                    | +--------+        |          |           | +--------+            |                                 
                    +-------------------+          |           +-----------------------+     
```

> A real use case would be: The hub cluster is installed in a **private network** while the managed cluster is in a public cloud. right now, the managed cluster cannot connect to the hub cluster. **It also means deployments on managed cluster can not apply any resources to the hub cluster directly**.

Author: @clyang82 @qiujian16   Hello, we have added a real use case here, please have a review and see whether it's clear for you. Thanks~

Reviewer 2: quick question, in the above case. Is there any reason that the cluster-manager controllers need to be in the management cluster?

Author: That's a good point! The controller could be running in the hub cluster too in the above case.
In practice, I got a case that the management cluster is super powerful, and the customer wants to leverage it for serval hub clusters, therefore hub clusters could be very light-weigh.
